Avoid AggregateException in TaskExtensions.WaitAndGetResult

svick opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/20884.

This could theoretically ignore some exceptions, if the `Task` contains more than one. But `await` has the same issue, so I think doing this is okay.

<s>I didn't find any tests for `WaitAndGetResult` or any relevant tests for `Formatter.Format`, so I didn't add any tests for this. If I should, what kind of test should I add?</s>

Pilchie commented:
@sharwell @CyrusNajmabadi want to give this a glance?

CyrusNajmabadi commented:
If Stephen is good with it, i'm good :)

CyrusNajmabadi commented:
Actually, scratch that.  I'm actually confused.  This is what MSDN says about ```Task<T>.Wait(CancellationToken)```

https://msdn.microsoft.com/en-us/library/dd321315(v=vs.110).aspx


| Exception | Condition |
|-|-|
| OperationCanceledException | The cancellationToken was canceled. |
| ObjectDisposedException | The task has been disposed. |
| AggregateException | The task was canceled. The AggregateException.InnerExceptions collection contains a TaskCanceledException object.-or-An exception was thrown during the execution of the task. The AggregateException.InnerExceptionscollection contains information about the exception or exceptions. |

CyrusNajmabadi commented:
So isn't it the Task.Wait that is actually causing the Aggregate exception to be thrown?

svick commented:
@CyrusNajmabadi You're right. Let me come back with a version that actually fixes the issue, and this time with unit tests.

stephentoub commented:
> So isn't it the Task.Wait that is actually causing the Aggregate exception to be thrown?

Yeah, I didn't look at the code above what changed :)  In this case Result and GetAwaiter().GetResult() are identical in functionality, because the exception would emerge from the previous call, and it would only get to the result line if it completed successfully.

svick commented:
I have fixed the code and added tests, should be ready for another review.

I also had to fix `RenameTrackingTaggerProvider` code, which relied on `WaitAndGetResult` throwing `AggregateException`. Since all test are passing now, can I assume this didn't break any other code?

stephentoub commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.cs:
```diff
@@ -172,10 +172,10 @@ internal static bool WaitForIsRenamableIdentifier(Task<TriggerIdentifierKind> is
                 return isRenamableIdentifierTask.WaitAndGetResult(cancellationToken) != TriggerIdentifierKind.NotRenamable;
             }
-            catch (AggregateException e) when (e.InnerException is OperationCanceledException)
+            catch (TaskCanceledException)
```
Any reason you've switch from the broader OperationCanceledException to the more derived TaskCanceledException?

svick replied:
It's to differentiate between `WaitForIsRenamableIdentifier(canceledTask, CancellationToken.None)` and `WaitForIsRenamableIdentifier(unfinishedTask, new CancellationToken(canceled: true))`. [According to the tests](https://github.com/dotnet/roslyn/blob/9205e7f/src/EditorFeatures/Test/RenameTracking/RenameTrackingTaggerProviderTests.cs#L854-L857), the former should return `false`, while the latter should throw OCE. If I just caught OCE here, the second test would fail.

I don't know why it's designed this way, but I thought I should keep the existing behavior and this does that.

Though there is one case I (or the tests) didn't consider: if the task was faulted with OCE. In that case, I think the method should return `false`, so I have switched to catching OCE and checking its `CancellationToken`.

stephentoub commented on src/Workspaces/Core/Portable/Utilities/TaskExtensions.cs:
```diff
@@ -60,8 +60,10 @@ public static T WaitAndGetResult<T>(this Task<T> task, CancellationToken cancell
-            return task.Result;
+            // WaitAny waits for the task to complete with cancellation and without throwing;
+            // GetAwaiter().GetResult() is then used so that if the task failed, the exception is not wrapped in AggregateException
+            Task.WaitAny(new Task[] { task }, cancellationToken);
```
How important is it that this be very efficient?  This is penalizing with an extra array allocation what I assume is the common path (success) for avoiding an exception throw on what I assume is the uncommon path (failure).

If it matters, you could of course switch to just letting the exception get thrown and then catching it.  Or at the very least you could special case when the task is already completed (no idea how common that is in the usage of this method), surrounding this call with `if (!task.IsCompleted) { ... }`.

svick replied:
My guess is that blocking on async code (which is what this method seems to be used for) should be fairly rare, so a single allocation should be fine. And I would expect that calling it with a completed task should be rare.

But I don't actually know this for certain.

CyrusNajmabadi replied:
I'm with stephen.  I'd prefer to avoid the allocatoin.

svick replied:
@CyrusNajmabadi @stephentoub Ok, updated to catch+rethrow.

sharwell commented on src/EditorFeatures/Test/RenameTracking/RenameTrackingTaggerProviderTests.cs:
```diff
@@ -854,9 +854,15 @@ public void RenameTrackingDoesNotThrowAggregateException()
             Assert.Throws<OperationCanceledException>(() => RenameTrackingTaggerProvider.WaitForIsRenamableIdentifier(source.Task, new CancellationToken(canceled: true)));
             source.TrySetException(new Exception());
-            Assert.Throws<AggregateException>(() => RenameTrackingTaggerProvider.WaitForIsRenamableIdentifier(source.Task, CancellationToken.None));
+            Assert.Throws<Exception>(() => RenameTrackingTaggerProvider.WaitForIsRenamableIdentifier(source.Task, CancellationToken.None));
```
:bulb: It would be good to assert that the thrown exception is the same instance that was set on the previous line.

svick replied:
Something like this?

```c#
var thrownException = new Exception();
source.TrySetException(thrownException);
var caughtException = Assert.Throws<Exception>(() => RenameTrackingTaggerProvider.WaitForIsRenamableIdentifier(source.Task, CancellationToken.None));
Assert.Equal(thrownException, caughtException);
```

That seems too verbose to me, considering that `Assert.Throws` verifies that the type of the thrown exception is actually `Exception` and not a some derived type.

sharwell replied:
@svick Yes, except use `Assert.Same`. It's verbose, but serves as a very precise regression test, much stronger than just a type check.

CyrusNajmabadi commented:
Tagging @jasonmalinowski about the rename exceptions and what is important to preserve or not.
